src/datasets/arrow_dataset.py

```diff
@@ -3819,14 +3849,19 @@ def sort(
                     fingerprint=new_fingerprint, indices_cache_file_name=indices_cache_file_name
                 )
 
-        column_data = self._getitem(
-            column, format_type="pandas", format_columns=None, output_all_columns=False, format_kwargs=None
+        queried_dataset = query_table(
+            table=self._data,
+            key=slice(0, len(self)),
+            indices=self._indices if self._indices is not None else None,
         )
 
-        df_sorted = column_data.to_frame().sort_values(
-            column, ascending=not reverse, kind=kind, na_position=null_placement
-        )
-        indices = df_sorted.index.to_numpy()
+        sort_keys = [
+            (column_names[i], "ascending" if not reverse[i] else "descending") for i in range(len(column_names))
+        ]
+
+        indices = np.array(
+            pa.compute.sort_indices(queried_dataset, sort_keys=sort_keys, null_placement=null_placement)
+        ).astype(np.int64)
```

Comment:
We can avoid the NumPy conversion here and modify select slightly to support PyArrow arrays (call `to_numpy` on the indices if they are of type `pa.Array`/`pa.ChunkedArray`):

```suggestion
        indices = pc.sort_indices(sort_table, sort_keys=sort_keys, null_placement=null_placement)
```